mautic/mautic
# Long segment build time warning

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [ N ]
| New feature/enhancement? (use the a.x branch)      | [ N ]
| Deprecations?                          | [ N ]
| BC breaks? (use the c.x branch)        | [ N ]
| Automated tests included?              | [ Y ] 
| Related user documentation PR URL      | mautic/mautic-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation#... 
| Issue(s) addressed                     | Fixes #... 



#### Description:
This PR adds information about how long it takes to rebuild a segment and display a warning when it takes too long.
When you have a large contacts database it's easy to make really time consuming segment filters (e.g. 10min single segment build time). It would be great if Mautic users could check how much time it takes to build a segment.

Changes:
- Added info to the lead list details table called "Last built time" 
- Update build time every time the particular segment is being rebuilt for all contacts (not just for some). In the command for the community (when called without min, max and contact ID arguments) and in the finalize job for the cloud.
- Check if last_built_time is longer than 30 (configurable) seconds then add a warning next to the segment name in the segment list table with tooltip "This segment build time is %count% seconds.".

![image](https://user-images.githubusercontent.com/8580942/216556413-5ddbbb32-3d77-4d14-b97e-9467ad3427eb.png)
![image](https://user-images.githubusercontent.com/8580942/216556457-e304a0d6-ba0f-4f32-9482-7b1ff3372d80.png)



#### Steps to test this PR:


1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Create a new segment.
3. Add some contacts to your segment using a segment filter. E.g. you can create a email based segment filter to add all contacts that have email defined (not empty).
4. Make sure that last_built_time value in segment details has been updated.
5. Go to the Campaign Studio -> Configuration -> Segment Settings -> Show warning if segment build time is longer than X seconds to 0. This is not real life scenario, but even 1s segment build time would be hard to replicate on empty database. If you have larger DB you can try with higher value.
6. Go the segments view page /s/segments. Make sure that you see a warning next to the segment name. If you hover the cursor over it you should see the following text: "This segment build time is %count% seconds."
